app/services/pdf_rendering_service.rb

```diff
@@ -0,0 +1,121 @@
+class PdfRenderingService
+  class Error < StandardError; end
+  class MissingChromeBinaryError < Error; end
+  class ChromeTimeoutError < Error; end
+
+  # Create a place for us to work under the Rails `tmp/` dir
+  TMP_DIR_PATH = Rails.root.join("tmp", "pdf_rendering").to_s.freeze
+
+  # Set a maximum amount of time we will allow Chrome to attempt to render the
+  # PDF
+  PDF_CONVERSION_TIMEOUT_SECONDS = 30 # seconds
+
+  attr_reader :pdf, :html
+
+  def initialize(from_html:)
+    @html = from_html
+
+    # We create some temporary files and we want to use a consistent unique
+    # prefix for their names because it makes debugging easier if we can easily
+    # associate the input HTML and the output PDF
+    @unique_id = calculate_unique_id
+
+    # @pdf is filled in by the #render method
+    @pdf = nil
+  end
+
+  def render
+    # mutate @html in place (for performance reasons)
+    mutate_html_by_inserting_base_tag
+
+    # Create our temp dir if it does not already exist
+    create_tmp_dir_if_required
+
+    # Do some gardening on the temp dir to stop it filling the disk. We choose
+    # to do this inline because, although it does impact the performance of
+    # generating PDFs, that is not a big deal and having this code manage its
+    # own temp dir is much easier than coordinating background jobs to manage
+    # it.
+    TempDirJanitorService.new(tmp_dir_path: TMP_DIR_PATH).remove_old_files
+
+    # Create some empty temporary files
+    pdf_path = create_empty_pdf_file
+    html_path = create_empty_html_file
+
+    # Write the HTML we received to a file for Chrome to consume (Chrome does
+    # not support rendering from STDIN)
+    write_to_file(html_path, @html)
+
+    # Convert the HTML file into a PDF and store it in the given pdf_path
+    render_as_pdf(input_html_path: html_path, output_pdf_path: pdf_path)
+
+    # Finally we build the RenderedPdf instance which represents the results of
+    # our work to the rest of the system
+    @pdf = RenderedPdf.new(file_path: pdf_path)
+  end
+
+  private
+
+  def write_to_file(path, contents)
+    File.write(path, contents)
+  end
+
+  def calculate_unique_id
+    "#{Time.zone.now.to_i}-#{SecureRandom.hex(10)}"
+  end
+
+  def create_empty_html_file
+    path = File.join(TMP_DIR_PATH, "#{@unique_id}-input.html")
+    FileUtils.touch(path)
+    path
+  end
+
+  def create_empty_pdf_file
+    path = File.join(TMP_DIR_PATH, "#{@unique_id}-output.pdf")
+    FileUtils.touch(path)
+    path
+  end
+
+  def render_as_pdf(input_html_path:, output_pdf_path:)
+    cmd = "node #{Rails.root.join("bin", "render-pdf.js")} #{input_html_path} #{output_pdf_path}"
+    Rails.logger.info(cmd)
+
+    error_msg = "Chrome did not complete the PDF conversion within the #{PDF_CONVERSION_TIMEOUT_SECONDS} second timeout"
```

Comment:
Are you certain that the browser is always Chrome?

Reply:
While this does function in other browsers, it uses chrome for the pdf rendering and I think this message is just for the logs.